File: go/sbv2/rpc.go
package sbv2

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"

	rt "sb/go/sb/v2"
)

type RpcErrCode int

const (
	RpcOk        RpcErrCode = 200
	RpcNoConn    RpcErrCode = 0
	RpcTimeout   RpcErrCode = 408
	RpcReqErr    RpcErrCode = 400
	RpcRespErr   RpcErrCode = 500
	RpcNotAuth   RpcErrCode = 401
	RpcNotExist  RpcErrCode = 404
	defaultMaxRespBytes int64 = 4 * 1024 * 1024
)

type Client struct {
	BaseURL string
	HTTP    *http.Client
	Timeout time.Duration
	Retries int
	MaxRespBytes int64
	headers map[string]string
	headersMu sync.RWMutex
}

func NewClient(baseURL string) *Client {
	baseURL = strings.TrimRight(baseURL, "/")
	return &Client{
		BaseURL: baseURL,
		HTTP:    &http.Client{},
		Timeout: 5 * time.Second,
		Retries: 3,
		MaxRespBytes: defaultMaxRespBytes,
		headers: make(map[string]string),
	}
}

func SetClientHeader(c *Client, key, value string) {
	if c == nil { return }
	c.headersMu.Lock()
	defer c.headersMu.Unlock()
	if c.headers == nil { c.headers = make(map[string]string) }
	c.headers[key] = value
}
func GetClientHeader(c *Client, key string) string {
	if c == nil { return "" }
	c.headersMu.RLock()
	defer c.headersMu.RUnlock()
	return c.headers[key]
}
func RemoveClientHeader(c *Client, key string) {
	if c == nil { return }
	c.headersMu.Lock()
	defer c.headersMu.Unlock()
	delete(c.headers, key)
}

func SetClientAuthorization(c *Client, token string) { SetClientHeader(c, "Authorization", "Bearer "+token) }
func GetClientAuthorization(c *Client) string        { return GetClientHeader(c, "Authorization") }
func RemoveClientAuthorization(c *Client)            { RemoveClientHeader(c, "Authorization") }
func IsClientAuthorized(c *Client) bool              { return GetClientAuthorization(c) != "" }

func isTimeout(err error) bool {
	if err == nil { return false }
	if err == context.DeadlineExceeded { return true }
	if netErr, ok := err.(interface{ Timeout() bool }); ok && netErr.Timeout() { return true }
	return false
}

func doClient(c *Client, ctx context.Context, path string, body []byte) ([]byte, RpcErrCode) {
	if c == nil { return nil, RpcNoConn }
	if ctx == nil { ctx = context.Background() }
	httpClient := c.HTTP
	if httpClient == nil { httpClient = &http.Client{} }
	maxRetries := c.Retries
	if maxRetries < 0 { maxRetries = 0 }
	maxRespBytes := c.MaxRespBytes
	if maxRespBytes <= 0 { maxRespBytes = defaultMaxRespBytes }
	var resp *http.Response
	var err error

	for i := 0; i <= maxRetries; i++ {
		if i > 0 {
			timer := time.NewTimer(time.Duration(i) * time.Second)
			select {
			case <-ctx.Done():
				timer.Stop()
				return nil, RpcTimeout
			case <-timer.C:
			}
		}
		reqCtx := ctx
		cancel := func() {}
		if c.Timeout > 0 {
			reqCtx, cancel = context.WithTimeout(ctx, c.Timeout)
		}
		req, reqErr := http.NewRequestWithContext(reqCtx, "POST", c.BaseURL+path, bytes.NewReader(body))
		if reqErr != nil {
			cancel()
			return nil, RpcNoConn
		}
		c.headersMu.RLock()
		for k, v := range c.headers {
			req.Header.Set(k, v)
		}
		c.headersMu.RUnlock()
		if req.Header.Get("Content-Type") == "" {
			req.Header.Set("Content-Type", "application/octet-stream")
		}
		resp, err = httpClient.Do(req)
		cancel()
		if err != nil {
			if isTimeout(err) && i < maxRetries { continue }
			if isTimeout(err) { return nil, RpcTimeout }
			return nil, RpcNoConn
		}
		if resp.StatusCode == http.StatusRequestTimeout && i < maxRetries {
			resp.Body.Close()
			continue
		}
		break
	}
	if resp == nil { return nil, RpcNoConn }
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK { return nil, RpcErrCode(resp.StatusCode) }
	readLimit := maxRespBytes
	if readLimit < (1<<63 - 1) { readLimit++ }
	b, readErr := io.ReadAll(io.LimitReader(resp.Body, readLimit))
	if readErr != nil { return nil, RpcRespErr }
	if readLimit > maxRespBytes && int64(len(b)) > maxRespBytes { return nil, RpcRespErr }
	return b, RpcOk
}

// CallUserGetAbc 获取用户的id
func CallUserGetAbc(c *Client, ctx context.Context) (result OrderStatus, errCode RpcErrCode) {
	var buf bytes.Buffer

	body, status := doClient(c, ctx, "/user.get_abc", buf.Bytes())
	if status != RpcOk {
		return result, status
	}
	respBuf := bytes.NewBuffer(body)
	{
		value, err := rt.GetU8(respBuf)
		if err != nil { return result, RpcRespErr }
		item := OrderStatus(value)
		if !isOrderStatus(item) { return result, RpcRespErr }
		result = item
	}
	if respBuf.Len() != 0 { return result, RpcRespErr }
	return result, status
}

// CallUserGetAbcd 获取abcd
func CallUserGetAbcd(c *Client, ctx context.Context, page uint8, size uint8) (result OrderStatus, errCode RpcErrCode) {
	var buf bytes.Buffer
	{
		if err := rt.SetU8(&buf, page); err != nil { return result, RpcReqErr }
	}
	{
		if err := rt.SetU8(&buf, size); err != nil { return result, RpcReqErr }
	}

	body, status := doClient(c, ctx, "/user.get_abcd", buf.Bytes())
	if status != RpcOk {
		return result, status
	}
	respBuf := bytes.NewBuffer(body)
	{
		value, err := rt.GetU8(respBuf)
		if err != nil { return result, RpcRespErr }
		item := OrderStatus(value)
		if !isOrderStatus(item) { return result, RpcRespErr }
		result = item
	}
	if respBuf.Len() != 0 { return result, RpcRespErr }
	return result, status
}

// CallUserSetSimInfo 设置sim信息
// 无返回值
func CallUserSetSimInfo(c *Client, ctx context.Context, info *SimInfo) (errCode RpcErrCode) {
	var buf bytes.Buffer
	{
		if err := SetSimInfo(&buf, info); err != nil { return RpcReqErr }
	}

	body, status := doClient(c, ctx, "/user.set_sim_info", buf.Bytes())
	if status != RpcOk {
		return status
	}
	if len(body) != 0 { return RpcRespErr }
	return status
}

// CallGetCount 获取数量
func CallGetCount(c *Client, ctx context.Context, page uint8) (result uint8, errCode RpcErrCode) {
	var buf bytes.Buffer
	{
		if err := rt.SetU8(&buf, page); err != nil { return result, RpcReqErr }
	}

	body, status := doClient(c, ctx, "/get_count", buf.Bytes())
	if status != RpcOk {
		return result, status
	}
	respBuf := bytes.NewBuffer(body)
	{
		value, err := rt.GetU8(respBuf)
		if err != nil { return result, RpcRespErr }
		result = value
	}
	if respBuf.Len() != 0 { return result, RpcRespErr }
	return result, status
}

// CallGetBin 获取bin
func CallGetBin(c *Client, ctx context.Context, page uint8) (result []byte, errCode RpcErrCode) {
	var buf bytes.Buffer
	{
		if err := rt.SetU8(&buf, page); err != nil { return result, RpcReqErr }
	}

	body, status := doClient(c, ctx, "/get_bin", buf.Bytes())
	if status != RpcOk {
		return result, status
	}
	respBuf := bytes.NewBuffer(body)
	{
		state, err := rt.GetU8(respBuf)
		if err != nil { return result, RpcRespErr }
		value, err := rt.GetBinCompactInto(respBuf, state, nil)
		if err != nil { return result, RpcRespErr }
		result = value
	}
	if respBuf.Len() != 0 { return result, RpcRespErr }
	return result, status
}

File: fixtures/ts/sb/rpc.ts
import * as rt from "./runtime_core"
import { DefaultOrderStatus, IsAssignableOrderStatus, IsOrderStatus, NormalizeOrderStatus, OrderStatus, getOrderStatusListBody, setOrderStatusListBody } from "./enum"
import { SimInfo, getSimInfoListBody, newSimInfo, readSimInfo, setSimInfo, setSimInfoListBody } from "./struct_sim_info"

const defaultMaxRespBytes = 4 * 1024 * 1024;
const maxSafeRespBytes = Number.MAX_SAFE_INTEGER;
const maxTimeoutMs = 2147483647;

export enum RpcErrCode {
    Ok = 200,
    NoConn = 0,
    Timeout = 408,
    ReqErr = 400,
    RespErr = 500,
    NotAuth = 401,
    NotExist = 404,
}

export type RpcStatus = RpcErrCode | number;

export interface RpcConfig {
    host: string;
    headers?: Record<string, string>;
    timeout?: number;
    retries?: number;
    enableRetries?: boolean;
    maxRespBytes?: number;
}

export interface RpcClient {
    setHeader: (key: string, value: string) => void;
    getHeader: (key: string) => string | undefined;
    removeHeader: (key: string) => void;
    setAuthorization: (token: string) => void;
    getAuthorization: () => string | undefined;
    removeAuthorization: () => void;
    isAuthorized: () => boolean;
    userGetAbc: () => Promise<[OrderStatus, RpcStatus]>;
    userGetAbcd: (page: number, size: number) => Promise<[OrderStatus, RpcStatus]>;
    userSetSimInfo: (info: SimInfo) => Promise<RpcStatus>;
    getCount: (page: number) => Promise<[number, RpcStatus]>;
    getBin: (page: number) => Promise<[Uint8Array, RpcStatus]>;
}

type _RpcClientState = RpcClient & {
    config: RpcConfig;
    headers: Record<string, string>;
    timeout: number;
    retries: number;
    enableRetries: boolean;
    maxRespBytes: number;
    _fetch: (path: string, body: Uint8Array) => Promise<[Uint8Array | null, RpcStatus]>;
};
type RpcClientCtor = new (config: RpcConfig) => RpcClient;

function _RpcClientCtor(this: _RpcClientState, config: RpcConfig): void {
    this.config = { ...config, host: config.host.replace(/\/+$/, "") };
    this.headers = config.headers ? { ...config.headers } : {};
    const cfgTimeout = config.timeout;
    this.timeout = cfgTimeout !== undefined && Number.isFinite(cfgTimeout) && cfgTimeout >= 0 ? Math.min(Math.floor(cfgTimeout), maxTimeoutMs) : 5000;
    const cfgRetries = config.retries;
    this.retries = cfgRetries !== undefined && Number.isFinite(cfgRetries) && cfgRetries >= 0 ? Math.floor(cfgRetries) : 3;
    this.enableRetries = config.enableRetries === true;
    const cfgMaxRespBytes = config.maxRespBytes;
    this.maxRespBytes = cfgMaxRespBytes !== undefined && Number.isFinite(cfgMaxRespBytes) && cfgMaxRespBytes > 0 ? Math.min(Math.floor(cfgMaxRespBytes), maxSafeRespBytes) : defaultMaxRespBytes;
}
const _rpcClientProto = _RpcClientCtor.prototype as _RpcClientState;

_rpcClientProto.setHeader = function(this: _RpcClientState, key: string, value: string): void { this.headers[key] = value; };
_rpcClientProto.getHeader = function(this: _RpcClientState, key: string): string | undefined { return this.headers[key]; };
_rpcClientProto.removeHeader = function(this: _RpcClientState, key: string): void { delete this.headers[key]; };

_rpcClientProto.setAuthorization = function(this: _RpcClientState, token: string): void { this.setHeader("Authorization", `Bearer ${token}`); };
_rpcClientProto.getAuthorization = function(this: _RpcClientState): string | undefined { return this.getHeader("Authorization"); };
_rpcClientProto.removeAuthorization = function(this: _RpcClientState): void { this.removeHeader("Authorization"); };
_rpcClientProto.isAuthorized = function(this: _RpcClientState): boolean { return !!this.getAuthorization(); };

async function readResponseBytes(res: Response, maxRespBytes: number): Promise<[Uint8Array | null, RpcErrCode | null]> {
    const contentLength = res.headers.get("content-length");
    if (contentLength !== null) {
        const size = Number(contentLength);
        if (Number.isFinite(size) && size > maxRespBytes) return [null, RpcErrCode.RespErr];
    }

    if (res.body !== null) {
        const reader = res.body.getReader();
        const chunks: Uint8Array[] = [];
        let total = 0;
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            if (value === undefined) continue;
            total += value.byteLength;
            if (total > maxRespBytes) {
                try {
                    await reader.cancel();
                } catch {
                }
                return [null, RpcErrCode.RespErr];
            }
            chunks.push(value);
        }

        const bytes = new Uint8Array(total);
        let offset = 0;
        for (const chunk of chunks) {
            bytes.set(chunk, offset);
            offset += chunk.byteLength;
        }
        return [bytes, null];
    }

    const bytes = new Uint8Array(await res.arrayBuffer());
    if (bytes.byteLength > maxRespBytes) return [null, RpcErrCode.RespErr];
    return [bytes, null];
}

_rpcClientProto._fetch = async function(this: _RpcClientState, path: string, body: Uint8Array): Promise<[Uint8Array | null, RpcStatus]> {
    const maxRetries = this.enableRetries ? this.retries : 0;
    let lastStatus = RpcErrCode.NoConn;
    for (let i = 0; i <= maxRetries; i++) {
        if (i > 0) await new Promise((res) => setTimeout(res, i * 1000));
        const controller = new AbortController();
        let timeoutId: ReturnType<typeof setTimeout> | null = null;
        if (this.timeout > 0) timeoutId = setTimeout(() => controller.abort(), this.timeout);
        try {
            const res = await fetch(`${this.config.host}/${path}`, {
                method: "POST",
                headers: { "Content-Type": "application/octet-stream", ...this.headers },
                body: body as any,
                signal: controller.signal,
            });
            if (res.ok) {
                const [bytes, readErr] = await readResponseBytes(res, this.maxRespBytes);
                if (readErr !== null) return [null, readErr];
                return [bytes, RpcErrCode.Ok];
            }
            lastStatus = res.status;
            if (res.status === 408 && i < maxRetries) continue;
            return [null, res.status];
        } catch (e: any) {
            lastStatus = e && e.name === "AbortError" ? RpcErrCode.Timeout : RpcErrCode.NoConn;
            if (i < maxRetries) continue;
        } finally {
            if (timeoutId !== null) clearTimeout(timeoutId);
        }
    }
    return [null, lastStatus];
};

// 获取用户的id
_rpcClientProto.userGetAbc = async function(this: _RpcClientState, ): Promise<[OrderStatus, RpcStatus]> {
    const buf = new rt.Buffer();
    const [bytes, status] = await this._fetch("user.get_abc", buf.bytes);
    if (status !== RpcErrCode.Ok || bytes === null) return [DefaultOrderStatus(), status];
    const respBuf = new rt.Buffer(bytes);
    let result = DefaultOrderStatus() as any;
        {
            const [value, err] = rt.getU8(respBuf);
            if (err !== null) return [DefaultOrderStatus(), RpcErrCode.RespErr];
            const item = value as OrderStatus;
            if (!IsOrderStatus(item)) return [DefaultOrderStatus(), RpcErrCode.RespErr];
            result = item as any;
        }
    if (respBuf.len !== 0) return [DefaultOrderStatus(), RpcErrCode.RespErr];
    return [result as any, RpcErrCode.Ok];
};

// 获取abcd
_rpcClientProto.userGetAbcd = async function(this: _RpcClientState, page: number, size: number): Promise<[OrderStatus, RpcStatus]> {
    const buf = new rt.Buffer();
        {
            const encodeErr = rt.setU8(buf, page as any);
            if (encodeErr !== null) return [DefaultOrderStatus(), RpcErrCode.ReqErr];
        }
        {
            const encodeErr = rt.setU8(buf, size as any);
            if (encodeErr !== null) return [DefaultOrderStatus(), RpcErrCode.ReqErr];
        }
    const [bytes, status] = await this._fetch("user.get_abcd", buf.bytes);
    if (status !== RpcErrCode.Ok || bytes === null) return [DefaultOrderStatus(), status];
    const respBuf = new rt.Buffer(bytes);
    let result = DefaultOrderStatus() as any;
        {
            const [value, err] = rt.getU8(respBuf);
            if (err !== null) return [DefaultOrderStatus(), RpcErrCode.RespErr];
            const item = value as OrderStatus;
            if (!IsOrderStatus(item)) return [DefaultOrderStatus(), RpcErrCode.RespErr];
            result = item as any;
        }
    if (respBuf.len !== 0) return [DefaultOrderStatus(), RpcErrCode.RespErr];
    return [result as any, RpcErrCode.Ok];
};

// 设置sim信息
// 无返回值
_rpcClientProto.userSetSimInfo = async function(this: _RpcClientState, info: SimInfo): Promise<RpcStatus> {
    const buf = new rt.Buffer();
        {
            const encodeErr = setSimInfo(buf, info);
            if (encodeErr !== undefined) return RpcErrCode.ReqErr;
        }
    const [bytes, status] = await this._fetch("user.set_sim_info", buf.bytes);
    if (status !== RpcErrCode.Ok || bytes === null) return status;
    if (bytes.byteLength !== 0) return RpcErrCode.RespErr;
    return RpcErrCode.Ok;
};

// 获取数量
_rpcClientProto.getCount = async function(this: _RpcClientState, page: number): Promise<[number, RpcStatus]> {
    const buf = new rt.Buffer();
        {
            const encodeErr = rt.setU8(buf, page as any);
            if (encodeErr !== null) return [0, RpcErrCode.ReqErr];
        }
    const [bytes, status] = await this._fetch("get_count", buf.bytes);
    if (status !== RpcErrCode.Ok || bytes === null) return [0, status];
    const respBuf = new rt.Buffer(bytes);
    let result = 0 as any;
        {
            const [value, err] = rt.getU8(respBuf);
            if (err !== null) return [0, RpcErrCode.RespErr];
            result = value as any;
        }
    if (respBuf.len !== 0) return [0, RpcErrCode.RespErr];
    return [result as any, RpcErrCode.Ok];
};

// 获取bin
_rpcClientProto.getBin = async function(this: _RpcClientState, page: number): Promise<[Uint8Array, RpcStatus]> {
    const buf = new rt.Buffer();
        {
            const encodeErr = rt.setU8(buf, page as any);
            if (encodeErr !== null) return [new Uint8Array(0), RpcErrCode.ReqErr];
        }
    const [bytes, status] = await this._fetch("get_bin", buf.bytes);
    if (status !== RpcErrCode.Ok || bytes === null) return [new Uint8Array(0), status];
    const respBuf = new rt.Buffer(bytes);
    let result = new Uint8Array(0) as any;
        {
            const [state, errState] = rt.getU8(respBuf);
            if (errState !== null) return [new Uint8Array(0), RpcErrCode.RespErr];
            const [value, err] = rt.getBin(respBuf, state);
            if (err !== null) return [new Uint8Array(0), RpcErrCode.RespErr];
            result = value as any;
        }
    if (respBuf.len !== 0) return [new Uint8Array(0), RpcErrCode.RespErr];
    return [result as any, RpcErrCode.Ok];
};

export const RpcClient = _RpcClientCtor as unknown as RpcClientCtor;
